libopae++/src/properties.cpp

```diff
@@ -69,6 +69,196 @@ properties::properties()
+      function(&props_, fpgaPropertiesGetFunction, fpgaPropertiesSetFunction),
+      socket_id(&props_, fpgaPropertiesGetSocketID, fpgaPropertiesSetSocketID),
+      num_slots(&props_, fpgaPropertiesGetNumSlots, fpgaPropertiesSetNumSlots),
+      bbs_id(&props_, fpgaPropertiesGetBBSID, fpgaPropertiesSetBBSID),
+      bbs_version(&props_, fpgaPropertiesGetBBSVersion,
+                  fpgaPropertiesSetBBSVersion),
+      vendor_id(&props_, fpgaPropertiesGetVendorID, fpgaPropertiesSetVendorID),
+      model(&props_, fpgaPropertiesGetModel, fpgaPropertiesSetModel),
+      local_memory_size(&props_, fpgaPropertiesGetLocalMemorySize,
+                        fpgaPropertiesSetLocalMemorySize),
+      capabilities(&props_, fpgaPropertiesGetCapabilities,
+                   fpgaPropertiesSetCapabilities),
+      num_mmio(&props_, fpgaPropertiesGetNumMMIO, fpgaPropertiesSetNumMMIO),
+      num_interrupts(&props_, fpgaPropertiesGetNumInterrupts,
+                     fpgaPropertiesSetNumInterrupts),
+      accelerator_state(&props_, fpgaPropertiesGetAcceleratorState,
+                        fpgaPropertiesSetAcceleratorState),
+      object_id(&props_, fpgaPropertiesGetObjectID, fpgaPropertiesSetObjectID),
+      parent(&props_, fpgaPropertiesGetParent, fpgaPropertiesSetParent),
+      guid(&props_) {
+  auto res = fpgaCloneProperties(p.props_, &props_);
+  if (res != FPGA_OK) {
+    log_.error() << "fpgaCloneProperties() failed with (" << res
+                 << ") " << fpgaErrStr(res);
+    throw except(res, OPAECXX_HERE);
+  }
+  // make sure that we have a cached copy of
+  // everything that p has cached.
+  if (p.type.is_set()) {
+    type.operator fpga_objtype();
+  }
+  if (p.bus.is_set()) {
+    bus.operator uint8_t();
+  }
+  if (p.device.is_set()) {
+    device.operator uint8_t();
+  }
+  if (p.function.is_set()) {
+    function.operator uint8_t();
+  }
+  if (p.socket_id.is_set()) {
+    socket_id.operator uint8_t();
+  }
+  if (p.num_slots.is_set()) {
+    num_slots.operator uint32_t();
+  }
+  if (p.bbs_id.is_set()) {
+    bbs_id.operator uint64_t();
+  }
+  if (p.bbs_version.is_set()) {
+    bbs_version.operator fpga_version();
+  }
+  if (p.vendor_id.is_set()) {
+    vendor_id.operator uint16_t();
+  }
+  if (p.model.is_set()) {
+    model.operator std::string();
+  }
+  if (p.local_memory_size.is_set()) {
+    local_memory_size.operator uint64_t();
+  }
+  if (p.capabilities.is_set()) {
+    capabilities.operator uint64_t();
+  }
+  if (p.num_mmio.is_set()) {
+    num_mmio.operator uint32_t();
+  }
+  if (p.num_interrupts.is_set()) {
+    num_interrupts.operator uint32_t();
+  }
+  if (p.accelerator_state.is_set()) {
+    accelerator_state.operator fpga_accelerator_state();
+  }
+  if (p.object_id.is_set()) {
+    object_id.operator uint64_t();
+  }
+  if (p.parent.is_set()) {
+    parent.operator fpga_token();
+  }
+  if (p.guid.is_set()) {
+    guid.operator uint8_t *();
+  }
+}
```

Comment:
Would it make sense to add a function called `pvalue<T>::update` that would call the getter to update the copy?

Reply:
Yeah, in fact the type cast operators should really be const. Using them in this way is nonintuitive.

Reply:
👍